source/common/conn_pool/conn_pool_base.cc

```diff
@@ -61,8 +82,8 @@ bool ConnPoolImplBase::shouldCreateNewConnection(float global_preconnect_ratio)
   //
   // If preconnect ratio is not set, it defaults to 1, and this simplifies to the
   // legacy value of pending_streams_.size() > connecting_stream_capacity_
-  return (pending_streams_.size() + num_active_streams_) * perUpstreamPreconnectRatio() >
-         (connecting_stream_capacity_ + num_active_streams_);
+  return shouldConnect(pending_streams_.size(), num_active_streams_, connecting_stream_capacity_,
+                       perUpstreamPreconnectRatio());
```

Comment:
Is the idea here basically to use either the global ratio or the per-upstream ratio to decide on whether to make a pre-connect for this host only? If so would it more clear to call this once and tack the max of the global ratio and the per-upstream ratio? I'm confused why we would pass true for anticipate above but not here?

Reply:
fundamentally there's two types of prefetching, local and global.  local is done every time there's a connection change, so will be adequately prefetched.  It doesn't need to anticipate as it's called after new streams are assigned.  global does need to anticipate, or incoming traffic to upstream A could never prefetch a connection for B (given the streams in B would be zero, it'd zero out the function).    So this function will only meaningfully call one of the two blocks.  I'll make it more clear with an if-else and add some more comments about the +1.